### PR TITLE
fix: solve problem that variable (of declaredType)  is not assigned with correct type

### PR DESCRIPTION
**Pull request by ltzmaxwell**
code like this won't work. when  `abs` is assigned like this: `abs = []Word{0}`, abs is assigned with a type of `[]Word`, thus `abs.add3()` does not work.

```
type nat3 []Word
type Word uint


func (n nat3) add3() bool {
	return true
}

func main() {
	var abs nat3
	abs = []Word{0}
	abs.add3()
}
```

another example, `abs` is a nested field:

```
type nat3 []Word
type Word uint

type Int3 struct {
	neg bool
	abs nat3
}

func (n nat3) add3() bool {
	return true
}

func main() {
	z := &Int3{
		neg: true,
		abs: []Word{0},
	}
	z.abs.add3()
}
```

The problem come from :  #651 .

The solution is to give proper type to variable with declaredType.

### CODE REVIEW

**Comment by ltzmaxwell** [comment]
Hi, @r3v4s , would you check this.

File: tests/files/method38gno [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package main
```
Comment (moul) [inline tests/files/method38gno:1]: invalid file name

**Comment by piux2** [comment]
@ltzmaxwell good findings! We are definitely in the right direction.

However, there are a lot of changes. Can you help describe the root cause and why the fix solved the problem?

I noticed a side effect in this test case.
examples/gno.land/r/demo/nft/z_0_filetest.gno

https://github.com/gnolang/gno/pull/674/files#diff-5a26a017ddff81d0ef1a36c2190cab1fedc6eac9cee7e404c62bc7a3bb73c303L54

     
    //             "T": {
    //             -    "@type": "/gno.PrimitiveType",
    //              -  "value": "16"
    //              +   "@type": "/gno.RefType",
    //              +   "ID": "gno.land/p/demo/grc/grc721.TokenID"
    //             },
    //             "V": {
    //                 "@type": "/gno.StringValue",
    //                 "value": "NFT#1"
    //             }
    //         }

     type NFToken struct {
	        owner    std.Address
	        approved std.Address
	        tokenID  grc721.TokenID
	        data     string
        }

Above is the "data" field metadata in an NFTToken instance. The value is "NFT#1"
Its Type should be   

    "@type": "/gno.PrimitiveType", "value": "16"

instead of 

    @type": "/gno.RefType", "ID": "gno.land/p/demo/grc/grc721.TokenID"

File: tests/files/method38gno (hunk 1 above)
Comment (ltzmaxwell) [reply tests/files/method38gno]: done

**Comment by ltzmaxwell** [comment]
Thank you @piux2 , it's actually a bug, I'm giving fix.

**Comment by ltzmaxwell** [comment]
seems be more reasonable to happen in preprocess.

**Comment by ltzmaxwell** [comment]
@peter7891 is ok for review

**Comment by jaekwon** [comment]
Do not merge, please see https://github.com/gnolang/gno/issues/651#issuecomment-1535451392

Also please add me as a reviewer for *all* gnolang vm changes!

**Comment by ltzmaxwell** [comment]
WIP, block.

**Comment by ltzmaxwell** [comment]
Hi, @jaekwon , I have another question,  currently the  code below would not work since n2 is not assigned with correct type from the assignment .

the question is whether we can fix it in preprocess stage?
In case like, `a, b = x, y.`, when traverse into `AssignStmt` node in preprocess, the `Rhs` can be  can be easily converted to the target type , but in case like `n1, n2 = x()` , seems we can not do the something without a list of  result Exprs of the `CallExpr`? 

Thank you!

```
package main

type nat []int

func (n nat) add() bool {
	return true
}

func x() (nat, []int) {
	a := nat{0}
	b := []int{1}
	return a, b
}

func main() {
	var n1, n2 nat

	n1, n2 = x()

	// n1.add()
	n2.add()

}
```

**Comment by jaekwon** [comment]
this is the solution we want: https://gist.github.com/piux2/49bada7be72186371e6684d007d3ed75?permalink_comment_id=4579166#gistcomment-4579166 links to https://github.com/gnolang/gno/compare/master...piux2:gno:fix_declaredtype_preprocess?expand=1 . 

The thread on piux2's thread (first link) has comments on improvements and what is necessary to make complete. 

To answer your exact question, yes we can and we will, but the simplest solution for now is to do the following:

```go
// pseudocode
// on LEAVE AssignStmt around 1574 for `case *CallExpr:`
    ...
    if dest type is interface type { 
        ...
    } else if leftType is exactly rightType {
        // perfect match, continue
    } else if checkType(leftType, rightType) {
        panic("a, b := x() implicit conversion not supported yet")
    } else {
        panic("type check error")
    }
```

Ray has the context to complete this. Then, you can work with Ray to complete the next PR which is to do the expansion only when necessary.

File: gnovm/pkg/gnolang/preprocess.go [hunk 2]
```diff
@@ -2364,7 +2379,7 @@ func convertConst(store Store, last BlockNode, cx *ConstExpr, t Type) {
 // Assert that xt can be assigned as dt (dest type).
 // If autoNative is true, a broad range of xt can match against
 // a target native dt type, if and only if dt is a native type.
-func checkType(xt Type, dt Type, autoNative bool) {
+func checkType(xt Type, dt Type, autoNative bool, conversionNeeded *bool) {
```
Comment (jaekwon) [inline gnovm/pkg/gnolang/preprocess.go:2382]: right, right. this would make sense (or it could just return the value as a bool, i think is simpler) to get the information out of the function. we will need this later for the full unroll solution (e.g. the second pr after ray's), but I suggest a return bool value instead.

File: gnovm/pkg/gnolang/preprocess.go [hunk 3]
```diff
@@ -1008,10 +1008,16 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 						}
 					} else if fv.PkgPath == uversePkgPath && fv.Name == "copy" {
 						if len(n.Args) == 2 {
+							args0T := evalStaticTypeOf(store, last, n.Args[0])
+							args1T := evalStaticTypeOf(store, last, n.Args[1])
+							if args0T.Elem().TypeID() != args1T.Elem().TypeID() {
+								panic(fmt.Sprintf(
+									"arguments to copy have different element types, want %s, got %s", args0T.Elem().TypeID(), args1T.Elem().TypeID()))
```
Comment (jaekwon) [inline gnovm/pkg/gnolang/preprocess.go:1015]: what prompted this? seems like this could/should be a separate PR for complete discussion. seems correct though.

**Review by jaekwon — CHANGES_REQUESTED** [review]
we can close this or @piux2 can merge here.

File: gnovm/pkg/gnolang/preprocess.go (hunk 3 above)
Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/preprocess.go:1015]: It's a fix of `copy`  function. According to the Go specification, "The core types of both arguments must be slices with identical element types". Our previous implementation lacked this necessary check.

I'll put it into another PR. Thank you~

**Comment by ltzmaxwell** [comment]
house keeping.
